### ISSUE TITLE
read_until: only works when no timeout is set

### ISSUE THREAD

**Issue by lHerzberger** [pyserial/pyserial#181]
Working on a raspberry pi 1 the following line does not work as soon as I set a timeout:
`callback = ser.read_until(terminator, 1000)`, where `terminator` is set to `b'\x03'`
as long as `ser.timeout` is set to `None`, my script reads all messages received over the serial port, but as soon as `ser.timeout` is set to anything other than `None` `callback` never contains more than 8 bytes.

**Comment by herm** [comment]
Have you tried calling ``read`` directly and checked if you run into a timeout there too. Or perhaps set your timeout to some very large number. 

``read_until`` works by calling ``read(1)`` repeatedly and does not care about the timeout at all.

**Comment by lHerzberger** [comment]
You're right it's not really a timeout thing (although `read_until` works fine for me as long as I don't set a timeout). Everything I've described below was tested with and without a timeout set (to 10 seconds mostly), so it's probably something else.

I've tried calling `read` now, but this only works when I specify the number of bytes to read (which I can't always predict). So `read()` stops reading after the first byte that's received and `read(10)` reads all 10 bytes that I'm sending in my test case.

I've also tried writing a loop that does pretty much the same as `read_until` (so it repeatedly calls `read(1)`) and checks if some terminator symbol was received. After receiving a maximum of 8 bytes it stops receiving bytes and the loop goes on forever, but never reading the last 2 bytes.
I then replaced `read(1)` with `read(ser.in_waiting)` the whole message is received within 2 iterations (`in_waiting=8` and then `in_waiting=2`). The problem is that this solution only works when I `print` something within every iteration (`in_waiting` or the received bytes), so it probably has something to do with the timing of the calls?

**Comment by jeffsilverm** [comment]
Herm, if I understand your comment of Nov 18th, 2016, does that mean that ```read_until``` might block forever if it is listening to something sending a stream of bytes that does *not* contain the terminator string?  

Now, if, for example, I had done something stupid such as wait for the "password" prompt instead of the "Password" prompt, then the current implementation works.  The server will send the "Password" prompt, and since I am waiting for "password", there will not be a match, and my application will wait until a ```read(1)``` call times out.  So that works.  But suppose I connected to something that is sending random bytes.  Then I might wait, if not forever at least for a long time before I get the match, and that only because sooner or later, a random series of bytes will the have the bytes 'p','a','s','s','w','o','r','d'.

So what I _think_ I want is a timeout on the read_until method itself.  Or is there another way to resolve this conundrum?

Many thanks,


Jeff

**Comment by herm** [comment]
Your description is correct. However read_until is fixed already. See #182.

**Comment by LeoAndGit** [comment]
I use read_until() to receive the message. Everytime I set the timeout=5 then send the message after read(1) is called. It can only return one byte once. I guess the timeout() in read should be set a little shorter than the timeout() in read_until().

**Comment by BobBaeck** [comment]
The expected string in read_until must be a byte array.  Then is everthing okay.
e.g.  response = s.read_until('> '.encode('utf-8')).decode('utf-8')
To get the respinse as string.
